lib/sms/sms.c

```diff
@@ -0,0 +1,322 @@
+
+/** @brief AT command to an ACK in PDU mode. */
+#define AT_SMS_PDU_ACK "AT+CNMA=1"
+
+/** @brief Start of AT notification for incoming SMS. */
+#define AT_SMS_NOTIFICATION "+CMT:"
+#define AT_SMS_NOTIFICATION_LEN (sizeof(AT_SMS_NOTIFICATION) - 1)
+
+static struct at_param_list resp_list;
+static char resp[AT_SMS_RESPONSE_MAX_LEN];
+
+/**
+ * @brief Indicates that the module has been successfully initialized
+ * and registered as an SMS client.
+ */
+static bool sms_client_registered;
+
+/** @brief SMS PDU data. */
+struct sms_data {
+	char *alpha;
+	u16_t length;
+	char *pdu;
+};
+
+/** @brief SMS event. */
+static struct sms_data cmt_rsp;
+
+struct sms_subscriber {
+	/* Listener user context. */
+	void *ctx;
+	/* Listener callback. */
+	sms_callback_t listener;
+};
+
+/** @brief List of subscribers. */
+static struct sms_subscriber subsribers[CONFIG_SMS_MAX_SUBSCRIBERS_CNT];
+
+/** @brief Check if the received AT notification is a +CMT event. */
+static bool sms_is_cmd_notification(const char *const at_notif)
+{
+	if (at_notif == NULL) {
+		return false;
+	}
+
+	if (strlen(at_notif) <= AT_SMS_NOTIFICATION_LEN) {
+		return false;
+	}
+
+	return (strncmp(at_notif, AT_SMS_NOTIFICATION,
+			AT_SMS_NOTIFICATION_LEN) == 0);
+}
+
+/** @brief Parse the +CMT unsolicited received message in PDU mode. */
+static int sms_cmt_notif_parse(const char *const buf)
+{
+	/* Parse the received message. */
+	int err = at_parser_max_params_from_str(buf, NULL, &resp_list,
+						AT_CMT_PARAMS_COUNT);
+
+	if (err != 0) {
+		LOG_ERR("Unable to parse CMT notification, err=%d", err);
+		return err;
+	}
+
+	/* Check the AT response format. */
+	if (at_params_valid_count_get(&resp_list) != AT_CMT_PARAMS_COUNT) {
+		LOG_ERR("Invalid CMT notification format");
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+/** @brief Save the SMS notification parameters. */
+static int sms_cmt_notif_save(void)
+{
+	if (cmt_rsp.alpha != NULL) {
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
+	(void)at_params_size_get(&resp_list, 1, &alpha_len);
+
+	cmt_rsp.alpha = k_malloc(alpha_len + 1);
+	if (cmt_rsp.alpha == NULL) {
+		return -ENOMEM;
+	}
+	(void)at_params_string_get(&resp_list, 1, cmt_rsp.alpha, &alpha_len);
+	cmt_rsp.alpha[alpha_len] = '\0';
+
+	/* Length field saved as number. */
+	(void)at_params_short_get(&resp_list, 2, &cmt_rsp.length);
+
+	/* Save PDU as a null-terminated String. */
+	size_t pdu_len;
+	(void)at_params_size_get(&resp_list, 3, &pdu_len);
+	cmt_rsp.pdu = k_malloc(pdu_len + 1);
+	if (cmt_rsp.pdu == NULL) {
+		return -ENOMEM;
+	}
+
+	(void)at_params_string_get(&resp_list, 3, cmt_rsp.pdu, &pdu_len);
+	cmt_rsp.pdu[pdu_len] = '\0';
+
+	return 0;
+}
+
+/** @brief Handler for AT responses and unsolicited events. */
+void sms_at_handler(void *context, char *at_notif)
+{
+	ARG_UNUSED(context);
+	LOG_DBG("AT notification '%s'", log_strdup(at_notif));
```

Comment:
I would move this down so that we only log something when it's an SMS,
AT notifications are logged somewhere else already.

Reply:
Yes, it can be enabled from the `at_cmd` module (https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/7e6228bb7e451ff19628636506b7caf3995ca1ec/drivers/at_cmd/at_cmd.c#L174).
I will remove lines 149 and 153 then.